Add support for sustain pedal (pianoroll and chromagram)

maezawa-akira opened the pull request:
I addressed issue #130 , and added a functionality to take into account sustain pedals, when extracting the piano-roll and the chromagram.

craffel commented on pretty_midi/instrument.py:
```diff
@@ -113,6 +125,25 @@ def get_piano_roll(self, fs=100, times=None):
+                elif is_pedal_on and not is_current_pedal_on:
+                    subpr = piano_roll[:, time_pedal_on:time_now]
+                    pedaled = np.minimum(subpr.cumsum(1),
+                                         subpr.max(1)[:, np.newaxis])
```
I'm a little confused by this operation.  Isn't `subpr.cumsum(1)` always less than `subpr.max(1)` given that subpr is positive (and it better be)? If you want to dress all of this code up with comments it wouldn't hurt!

craffel commented on pretty_midi/instrument.py:
```diff
@@ -84,6 +86,16 @@ def get_piano_roll(self, fs=100, times=None):
+            Default is False, which does nothing.  If True, then CC64 value
+            greater than pedal_threshold will be treated as pedal on and
+            the rest as pedal off.
+        pedal_threshold : Int
```
Is there any standard for this?  Like what does it mean for a MIDI file to have a sustain pedal CC value of 1?

maezawa-akira replied:
Not really.  It depends more on how a synthesizer is implemented.  For example, piano-specific synthesizers interpret CC value of 1 to be a very soft pedal (like a Yamaha Disklavier), while most software synthesizers treat them as a no-pedal.  
So it's really dependent on the context of the MIDI data and the use case of this library's user.

craffel replied:
Ok, interesting.  Along these lines, I think the most common user preference would be for sustain pedal to be taken into consideration (corresponding to `use_pedal=True` in your code).  For that reason I'd suggest we unify `use_pedal` and `pedal_threshold` into a single argument, `pedal_threshold`, which defaults to 64.  If `pedal_threshold is None`, then we don't use the sustain pedal. What do you think?

As an aside, as a nitpick, please list types in the docstring literally, e.g. bool and int instead of Boolean and Int.

maezawa-akira commented:
I have added documentation for readability.  Also, while documenting the code, I found a much simpler method for taking the pedal, so I modified the code as well.

maezawa-akira commented:
Hi, I modified the code so that it uses a single parameter pedal_threshold to determine if CC64 should be used ( 93b94c0 )

craffel commented:
Ok, final suggestion: Should it default to `64` instead of `None`?  I would think people would, by default, want sustain pedal to be taken into consideration, even though it breaks backwards compatibility (which we're allowed to do since we're still pre-v1!)

cghawthorne commented on pretty_midi/instrument.py:
```diff
@@ -113,6 +120,29 @@ def get_piano_roll(self, fs=100, times=None):
+            for cc in [_e for _e in self.control_changes
+                       if _e.number == CC_SUSTAIN_PEDAL]:
+                time_now = int(cc.time*fs)
+                is_current_pedal_on = (cc.value > pedal_threshold)
```
Very minor nit, >= 64 seems like a better threshold to me because at 64, you're in the upper half of values.

Likely doesn't matter at all because for this kind of sustain processing, we more or less expect it to be just on or off.

maezawa-akira replied:
Thanks for the good point! I fixed the code accordingly.

cghawthorne commented on pretty_midi/instrument.py:
```diff
@@ -113,6 +120,29 @@ def get_piano_roll(self, fs=100, times=None):
+                    subpr = piano_roll[:, time_pedal_on:time_now]
+
+                    # Take the running maximum
+                    pedaled = np.maximum.accumulate(subpr, axis=1)
```
I'm not totally sure this is the correct way to handle repeated notes with sustain. The way we did it on magenta was to extend each note within a sustain to either the beginning of the next note of the same pitch or the end of the sustain period, whichever happens first:

https://github.com/tensorflow/magenta/blob/master/magenta/music/sequences_lib.py#L972

The pianoroll equivalent of that would be to just retain whatever the most recent velocity was, not the maximum. My argument for using the most recent velocity would be that if you play a note with high velocity on a piano with sustain on, then wait a long time, then play the same note with low velocity, you won't hear a loud note, you'll hear a soft note.

That said, I could see it going the other way too (play a soft note immediately after a loud note), but I think retaining the most recent velocity better captures what's going on without making too many assumptions about the acoustic properties of the instrument.

What do you think?

maezawa-akira replied:
Hi, I totally understand your point, but I opted to taking the running maximum for three reasons.

1. This approach seemed faster.  When generating piano rolls, this approach seems faster than keeping track of the current note on events, especially when the piano roll w/o pedal messages are already given.  
2. This approach covers the majority of use-case of CC64 on musical instruments. Using the most recent velo is a good assumption for decaying sounds that are played with a long interval.  Considering that pedals are often depressed for only a few bars, and also since CC64 are also used extensively on non-decaying instruments like string pads and organs, I argue that taking the running maximum captures the majority of the use-cases of CC64.    
3. I think the two approaches are complementary ways of approximating the pedals, but the choice would not influence most use-cases of how piano-rolls are used, as binary indicator of note presence.  If knowing the actual loudness is important, then one should convolve the note-on messages with the ADSR envelope.  But this kind of implementation is too time-consuming to be valuable in most cases of interest of using the piano-rolls like AMT or informed source separation; the interest is in knowing the existence of notes and not the loudness.

cghawthorne replied:
Yup, I can see the arguments that way too. I'll leave it to @craffel to decide. :-)

craffel commented:
Merged, thank you so much @maezawa-akira !
